renter/proto/session.go

```diff
@@ -26,36 +26,47 @@ import (
 var (
 	// ErrInsufficientFunds is returned by various RPCs when the renter is
 	// unable to provide sufficient payment to the host.
-	ErrInsufficientFunds = errors.New("insufficient funds")
+	ErrInsufficientFunds = fmt.Errorf("insufficient funds")
 
 	// ErrInvalidMerkleProof is returned by various RPCs when the host supplies
 	// an invalid Merkle proof.
-	ErrInvalidMerkleProof = errors.New("host supplied invalid Merkle proof")
+	ErrInvalidMerkleProof = fmt.Errorf("host supplied invalid Merkle proof")
 
 	// ErrContractLocked is returned by the Lock RPC when the contract in
 	// question is already locked by another party. This is a transient error;
 	// the caller should retry later.
-	ErrContractLocked = errors.New("contract is locked by another party")
+	ErrContractLocked = fmt.Errorf("contract is locked by another party")
 
 	// ErrNoContractLocked is returned by RPCs that require a locked contract
 	// when no contract is locked.
-	ErrNoContractLocked = errors.New("no contract locked")
+	ErrNoContractLocked = fmt.Errorf("no contract locked")
 
 	// ErrContractFinalized is returned by the Lock RPC when the contract in
 	// question has reached its maximum revision number, meaning the contract
 	// can no longer be revised.
-	ErrContractFinalized = errors.New("contract cannot be revised further")
+	ErrContractFinalized = fmt.Errorf("contract cannot be revised further")
 )
 
 // wrapResponseErr formats RPC response errors nicely, wrapping them in either
 // readCtx or rejectCtx depending on whether we encountered an I/O error or the
 // host sent an explicit error message.
 func wrapResponseErr(err error, readCtx, rejectCtx string) error {
-	err = errors.Cause(err)
+	// the original implementation (pkg/errors.Wrap() returns nil on nil err, unlike fmt.Errorf
+	if err == nil {
+		return nil
+	}
+	// we need the lowest error in the stack - the origin/"cause"
+	for {
+		uerr := errors.Unwrap(err)
+		if uerr == nil {
+			break
+		}
+		err = uerr
+	}
```

Comment:
I also don't think this implementation is correct, as `errors.Cause()` doesn't mean the lowest error in the stack necessarily, an the package implementation returns the topmost error that implements the interface... Someone with a bigger brain than me may need to fix this (in context of this package).

Reply:
I think this is what `errors.As` is for:
```go
if errors.As(err, new(*renterhost.RPCError)) {
    return fmt.Errorf("%s: %w", rejectCtx, err)
}
```
(The double indirection is weird, but correct; `As` wants a pointer to a value that implements `error`, and in this case `renterhost.RPCError` has a pointer receiver. Furthermore, since we don't need to actually inspect the error, we don't need to declare a separate variable for it.)